### PR TITLE
Fixes in DynamoDB event polling

### PR DESCRIPTION
**Pull request by klizhentas**
* Add resest for buffers to close watchers
and reset buffer the state.
* Add reconnect logic to DynamoDB
* Add tests for cache watchers, make sure
the errors of the cache internal watcher propagate to
external watchers.

### CODE REVIEW

File: lib/cache/cache_test.go [hunk 1]
```diff
@@ -267,6 +265,50 @@ func (s *CacheSuite) onlyRecentDisconnect(c *check.C) {
...
 }
 
+// TestWatchers tests watchers connected to the cache,
+// verifies that all watchers of the cache will be closed
+// if the underlying watcher to the target backend is closed
+func (s *CacheSuite) TestWatchers(c *check.C) {
+	p := s.newPackForAuth(c)
+	defer p.Close()
+
+	w, err := p.cache.NewWatcher(context.TODO(), services.Watch{Kinds: []services.WatchKind{
```
Comment (russjones) [inline lib/cache/cache_test.go]:
```go
w, err := p.cache.NewWatcher(context.TODO(), services.Watch{Kinds: []services.WatchKind{
	services.WatchKind{
		Kind: services.KindCertAuthority,
	},
}})
```

**Review by russjones — APPROVED** [review]

File: lib/backend/dynamo/dynamodbbk.go [hunk 2]
```diff
@@ -63,6 +64,8 @@ type DynamoConfig struct {
 	BufferSize int `json:"buffer_size,omitempty"`
 	// PollStreamPeriod is a polling period for event stream
 	PollStreamPeriod time.Duration `json:"poll_stream_period,omitempty"`
+	// RetryPeriod is a period between dynamo backend retries on failures
+	RetryPeriod time.Duration
```
Comment (r0mant) [inline lib/backend/dynamo/dynamodbbk.go]: Missing json tag.

File: lib/backend/dynamo/shards.go [hunk 3]
```diff
@@ -38,13 +39,41 @@ type shardEvent struct {
 }
 
 func (b *DynamoDBBackend) asyncPollStreams(ctx context.Context) error {
-	err := b.pollStreams(ctx)
+	retry, err := utils.NewLinear(utils.LinearConfig{
+		Step: b.RetryPeriod / 10,
```
Comment (r0mant) [inline lib/backend/dynamo/shards.go]: There's a chance that "step" will be 0 here (given low enough retry period). Not sure how this linear backoff will behave in this case - could potentially retry indefinitely?

File: lib/cache/cache.go [hunk 4]
```diff
@@ -329,6 +329,11 @@ func (c *Cache) update() {
 		err := c.fetchAndWatch(retry, time.After(c.ReloadPeriod))
 		if err != nil {
 			c.setCacheState(err)
+			// if cache is experiencing error,
+			// all watchers are out of sync, because
+			// cache has lost it's own watcher to the backend
```
Comment (r0mant) [inline lib/cache/cache.go]: Nit: it's -> its

File: lib/backend/buffer.go [hunk 5]
```diff
@@ -61,15 +61,27 @@ func NewCircularBuffer(ctx context.Context, size int) (*CircularBuffer, error) {
...
 	c.Lock()
 	defer c.Unlock()
 	for _, w := range c.watchers {
 		w.Close()
 	}
 	c.watchers = nil
+	c.start = -1
+	c.end = -1
+	c.size = 0
+	for i := 0; i < len(c.events); i++ {
```
[the remark this answers is not in the document]
Comment (klizhentas) [reply lib/backend/buffer.go]: no, it should not as the buffer is pre-allocated

File: lib/cache/cache_test.go (hunk 1 above)
Comment (klizhentas) [reply lib/cache/cache_test.go]: why extra boilerplate? Golan's latest addition makes it unnecessary

File: lib/backend/dynamo/shards.go (hunk 3 above)
Comment (klizhentas) [reply lib/backend/dynamo/shards.go]: I think it's possible if the retry period is less than 9ns what is not practical

**Pull request merged**